File: errors/haskell/src/Bevy.hs
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE IncoherentInstances #-}

module Bevy ( App
            , Res
            , Resource
            , IntoSystem
            , System
            , newApp
            , injectResource
            , addSystem
            , runApp
            ) where

newtype Res a = Res a

data JustSystem
data IsFunctionSystem

class Resource a
class SystemParam a
class SystemParamFunction a
class System a

-- NOTE: modeling the `Marker` as a closed type family
-- makes it easier for GHC to produce better diagnostics.
type family Marker a where
  Marker (a -> b) = IsFunctionSystem
  Marker a = JustSystem

class IntoSystem a
instance (Marker a ~ marker, IntoSystem' a marker) => IntoSystem a

class IntoSystem' a marker where {}
instance (System a) => IntoSystem' a JustSystem
instance (SystemParamFunction a) => IntoSystem' a IsFunctionSystem

instance (SystemParam a) =>
  SystemParamFunction (a -> out)
 
instance (SystemParam a, SystemParam b) =>
  SystemParamFunction (a -> b -> out)

instance (Resource a) => SystemParam (Res a)

data App = App

newApp :: () -> App
newApp = undefined

injectResource :: (Resource a) => App -> a -> App
injectResource = undefined

addSystem :: IntoSystem a => App -> a -> App
addSystem = undefined

runApp :: App -> IO ()
runApp = undefined

File: errors/scala/src/main/scala/Bevy.scala
package bevy

import scala.util.TupledFunction

trait IntoSystem[A, M]
trait SystemParamFunction[A]
trait System[A]
trait SystemParam[A]
trait Resource[A]

// Marker types
case class JustSystem()
case class IsFunctionSystem()

// Parameter wrapper
case class Res[R: Resource](r: R)


given [R: Resource]: SystemParam[Res[R]] with {}

given [T1: SystemParam, T2: SystemParam]:
  SystemParam[(T1, T2)] with {}

given [A: System]: IntoSystem[A, JustSystem] with {}
given [A: SystemParamFunction]: IntoSystem[A, IsFunctionSystem] with {}

// NOTE: produces an "ambiguous instace" because of the plain `Tuple`
// given [F, Args <: Tuple : SystemParam, Res](using TupledFunction[F, Args => Res]):
//   SystemParamFunction[F] with {}

given [Args: SystemParam, Res]:
  SystemParamFunction[Args => Res] with {}

case class App():
  def addSystem[S, M](s: S)(using intoSystem: IntoSystem[S, M]): Unit = ()
  def injectResource[R: Resource](r: R): Unit = ()
  def run(): Unit = println("Application Ran")
